rubix-core/src/main/java/com/qubole/rubix/core/NonLocalReadRequestChain.java

```diff
@@ -12,24 +12,115 @@
+import com.qubole.rubix.spi.CacheConfig;
+import com.qubole.rubix.spi.DataRead;
+import com.qubole.rubix.spi.RetryingBookkeeperClient;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.thrift.shaded.TException;
+
+import static com.qubole.rubix.spi.RetryingBookkeeperClient.createBookKeeperClient;
 
 /**
- * Created by qubole on 31/8/16.
+ * Created by sakshia on 31/8/16.
  */
-public class NonLocalReadRequestChain extends DirectReadRequestChain
+public class NonLocalReadRequestChain extends ReadRequestChain
 {
-    public NonLocalReadRequestChain(FSDataInputStream inputStream)
+    private final String filePath;
+    String remoteNodeName;
+    RetryingBookkeeperClient bookKeeperClient;
+    Configuration conf;
+    private boolean strictMode = false;
+    int totalRead = 0;
+    FileSystem fs;
+    FSDataInputStream inputStream = null;
+
+    private static final Log log = LogFactory.getLog(ReadRequestChain.class);
+
+    public NonLocalReadRequestChain(String remoteNodeName, Configuration conf, FileSystem fs, String remotePath)
     {
-        super(inputStream);
+        this.remoteNodeName = remoteNodeName;
+        this.fs = fs;
+        this.conf = conf;
+        this.strictMode = CacheConfig.isStrictMode(conf);
+        this.filePath = remotePath;
     }
 
     public ReadRequestChainStats getStats()
     {
         return new ReadRequestChainStats()
-                .setRemoteReads(requests)
                 .setNonLocalReads(requests)
                 .setRequestedRead(totalRead)
                 .setNonLocalDataRead(totalRead);
     }
+
+    @Override
+    public Integer call()
+            throws Exception
+    {
+        if (readRequests.size() == 0) {
+            return 0;
+        }
+
+        try {
+            this.bookKeeperClient = createBookKeeperClient(remoteNodeName, conf);
+        }
+        catch (Exception e) {
+            if (strictMode) {
+                throw Throwables.propagate(e);
+            }
+            log.warn("Could not create BookKeeper Client " + Throwables.getStackTraceAsString(e));
+            return directReadRequest(0);
+        }
+
+        for (ReadRequest readRequest : readRequests) {
+            int readLength = 0;
+            int lengthRemaining = readRequest.getActualReadLength();
+            int bufferLength = CacheConfig.getBufferSize(conf);
+
+            DataRead dataRead;
+
+            while (lengthRemaining > 0) {
+                if (lengthRemaining < bufferLength) {
+                    bufferLength = lengthRemaining;
+                }
+                try {
+                    dataRead = bookKeeperClient.readData(filePath, readRequest.getActualReadStart() + readLength, readRequest.getDestBufferOffset(), bufferLength);
```

Comment:
Also instead of passing `bufferLength`, we need to send `min(bufferLength, actualLength)`